File: bindings/numrs-js/examples/fraud_detection.js
const { Tensor, nn, optimizers: optim, NativeTensor, TrainerBuilder, Dataset } = require('../dist/native_loader');
const fs = require('fs');
const { Array: NumRsArray } = require('../dist/array');
// const data = require('../dist/data'); // Removing old data import

const { Linear, Sequential, ReLU, Sigmoid } = nn;

// Helper: Random Float in range
function randRange(min, max) {
    return Math.random() * (max - min) + min;
}

// 1. Data Generation
function getFraudData(n) {
    const xData = [];
    const yData = [];

    for (let i = 0; i < n; i++) {
        const amount = randRange(0.0, 1.0);
        const hour = randRange(0.0, 1.0);
        const risk = randRange(0.0, 1.0);

        // Simple fraud logic
        const isFraud = (risk > 0.8 && amount > 0.5) || (risk > 0.95);

        xData.push([amount, hour, risk]);
        yData.push([isFraud ? 1.0 : 0.0]);
    }
    return { x: xData, y: yData };
}

async function main() {
    console.log("🚀 Running Fraud Detection Example (MLP) with Trainer API...");

    // 2. Prepare Data
    console.log("📊 Generating Dataset...");
    const { x, y } = getFraudData(200);
    // Use Native Dataset (renamed to Dataset in exports)
    const dataset = new Dataset(x, y, 32);
    console.log(`   Created dataset with ${dataset.numBatches} batches.`);

    // 3. Define Model
    console.log("\n🧠 Defining Architecture...");
    const model = new Sequential();
    model.addLinear(new Linear(3, 16));
    model.addRelu(new ReLU());
    model.addLinear(new Linear(16, 8));
    model.addRelu(new ReLU());
    model.addLinear(new Linear(8, 1));
    model.addSigmoid(new Sigmoid());
    console.log("   Linear(3->16) -> ReLU -> Linear(16->8) -> ReLU -> Linear(8->1) -> Sigmoid");

    // 4. Training Setup via Trainer API
    const EPOCHS = 50;

    // Validation Data
    const valDataRaw = getFraudData(50);
    const valDataset = new Dataset(valDataRaw.x, valDataRaw.y, 32);

    console.log("\n🏋️  Training via Trainer...");
    const builder = new TrainerBuilder(model);
    builder.learningRate(0.01);

    // Using Adam and MSE
    const trainer = builder.build("adam", "mse");

    const history = trainer.fit(dataset, valDataset, EPOCHS, true);

    console.log("✓ Training Complete");

    // Log final metrics (napi converts to camelCase)
    const lastMetrics = history[history.length - 1];
    console.log(`   Final Train Loss: ${lastMetrics.trainLoss.toFixed(4)}`);
    if (lastMetrics.valLoss !== undefined) {
        console.log(`   Final Val Loss:   ${lastMetrics.valLoss.toFixed(4)}`);
    }

    // 6. Simulate Production Inference (Manual Forward Pass)
    console.log("\n🚀 Simulating Production Inference (Manual Forward)...");
    const cases = [
        { amt: 0.1, hr: 0.5, rsk: 0.1, label: "Legit" },  // Low risk
        { amt: 0.9, hr: 0.2, rsk: 0.96, label: "Fraud" }, // High risk > 0.95
        { amt: 0.6, hr: 0.8, rsk: 0.85, label: "Fraud" }, // Risk > 0.8 && Amt > 0.5
        { amt: 0.2, hr: 0.9, rsk: 0.85, label: "Legit" }, // High risk but Low Amt
    ];

    console.log("   ---------------------------------------------------------------");
    console.log("   | Amount | Hour   | Risk   | Prob     | Pred     | Match? |");
    console.log("   ---------------------------------------------------------------");

    for (const c of cases) {
        // NativeTensor expects Float32Array for data
        const inputDataCase = new Float32Array([c.amt, c.hr, c.rsk]);
        // NumRsArray wrapping Float32Array
        // Tensor wrapping NumRsArray
        const inputCase = new Tensor(new NumRsArray(inputDataCase, [1, 3]), [1, 3], false);

        const outCase = model.forward(inputCase);

        // outCase is Tensor. data getter returns NumRsArray. data getter returns Float32Array.
        // Or directly values() if bound? 
        // Let's assume .data.data access pattern from TS example.
        const probCase = outCase.data.data[0];
        const predStrCase = probCase > 0.5 ? "Fraud" : "Legit";
        const match = predStrCase === c.label ? "✅" : "❌";

        console.log(`   | ${c.amt.toFixed(2).padEnd(6)} | ${c.hr.toFixed(2).padEnd(6)} | ${c.rsk.toFixed(2).padEnd(6)} | ${probCase.toFixed(4).padEnd(8)} | ${predStrCase.padEnd(8)} |   ${match}    |`);
    }
    console.log("   ---------------------------------------------------------------");

    // 7. Export (Optional, mocking ONNX save via binding if available)
    if (model.saveOnnx) {
        console.log("\n💾 Exporting to ONNX...");
        try {
            const dummy_f32 = new Float32Array([0.0, 0.0, 0.0]);
            const dummy_input = new Tensor(new NumRsArray(dummy_f32, [1, 3]), [1, 3], false);
            model.saveOnnx(dummy_input, "fraud_model_trainer.onnx");
            console.log("   Model saved.");
            if (fs.existsSync("fraud_model_trainer.onnx")) {
                fs.unlinkSync("fraud_model_trainer.onnx");
                console.log("   Cleaned up.");
            }
        } catch (e) {
            console.log("   Export failed: " + e);
        }
    }
}

main().catch(console.error);

File: numrs-core/examples/fraud_detection.rs
use anyhow::Result;
use numrs::autograd::nn::{Linear, ReLU, Sequential, Sigmoid};
use numrs::autograd::{LossFunction, Module, Tensor};
use numrs::ops::model::{infer, load_onnx};
use numrs::ops::{add, log, mean, mul, neg, sub};
use numrs::{Array, Dataset, TrainerBuilder};
use rand::prelude::*;
use std::collections::HashMap;
use std::path::Path;

// =========================================================================================
// 1. DATA GENERATION
// =========================================================================================

/// Genera datos sintéticos para fraud detection
/// Features: [Amount, Hour, RiskScore]
/// Label: 1.0 (Fraud) | 0.0 (Legit)
fn get_fraud_data(n: usize) -> (Vec<Vec<f32>>, Vec<f32>) {
    let mut rng = thread_rng();
    let mut x = Vec::with_capacity(n);
    let mut y = Vec::with_capacity(n);

    for _ in 0..n {
        // Feature 1: Monto normalizado (0.0 = $0, 1.0 = $10,000)
        let amount: f32 = rng.gen_range(0.0..1.0);
        // Feature 2: Hora del día normalizada (0.0 = 00:00, 1.0 = 23:59)
        let hour: f32 = rng.gen_range(0.0..1.0);
        // Feature 3: Score de Riesgo externo (0.0 = Seguro, 1.0 = Peligroso)
        let risk: f32 = rng.gen_range(0.0..1.0);

        // LÓGICA DE NEGOCIO (Ground Truth):
        // 1. Si el riesgo es extremo (>0.95) -> FRAUDE SEGURO
        // 2. Si el riesgo es alto (>0.8) Y el monto es alto (>0.5) -> FRAUDE
        let is_fraud = (risk > 0.8 && amount > 0.5) || (risk > 0.95);

        let label = if is_fraud { 1.0 } else { 0.0 };

        x.push(vec![amount, hour, risk]);
        y.push(label);
    }
    (x, y)
}

// =========================================================================================
// 2. CUSTOM LOSS FUNCTION
// =========================================================================================

/// Binary Cross Entropy Loss personalizada
struct BCELoss;

impl LossFunction for BCELoss {
    fn compute(&self, preds: &Tensor, targets: &Tensor) -> Result<Tensor> {
        // preds: Probabilidades [0..1] (Output de Sigmoid)
        // targets: Etiquetas [0 o 1]

        // NumRs doesn't support automatic broadcasting yet, so we must manually
        // create constants with the same shape as inputs [BatchSize, 1].
        // 1. Estabilidad Numérica (Epsilon)
        // 1. Estabilidad Numérica (Epsilon)
        // Now using automatic broadcasting and f64->f32 casting!
        let epsilon = Tensor::new(Array::new(vec![1], vec![1e-7]), false);

        // preds_safe = preds + epsilon
        let preds_safe = preds.add(&epsilon)?;

        // 2. Primer Término: - y * log(p)
        let log_p = preds_safe.log()?;
        let term1 = targets.mul(&log_p)?; // y * log(p)

        // 3. Segundo Término: - (1-y) * log(1-p)
        // "1" Scalar Tensor
        let one = Tensor::new(Array::new(vec![1], vec![1.0]), false);

        // "-1" Scalar Tensor (for subtraction via add + mul(-1))
        let neg_one = Tensor::new(Array::new(vec![1], vec![-1.0]), false);

        // one_minus_y = 1 + (y * -1)
        let neg_targets = targets.mul(&neg_one)?;
        let one_minus_y = one.add(&neg_targets)?;

        // one_minus_p = 1 + (p * -1)
        let neg_preds = preds_safe.mul(&neg_one)?;
        let one_minus_p = one.add(&neg_preds)?;

        // Add epsilon to (1-p)
        let one_minus_p_safe = one_minus_p.add(&epsilon)?;
        let term2 = one_minus_y.mul(&one_minus_p_safe.log()?)?; // (1-y)*log(1-p)

        // 4. Loss = - mean(term1 + term2)
        //    Loss = mean(term1 + term2) * -1
        let total = term1.add(&term2)?;
        let mean_val = total.mean()?; // mean() is a method on Tensor
        mean_val.mul(&neg_one)
    }
}

// =========================================================================================
// MAIN
// =========================================================================================

fn main() -> Result<()> {
    // FORCE CPU BACKEND FOR EXAMPLE STABILITY
    //std::env::set_var("NUMRS_BACKEND", "cpu");

    println!("═══════════════════════════════════════════════════════════");
    println!("  🚨  NumRs Example: Fraud Detection (Tutorial 06)");
    println!("═══════════════════════════════════════════════════════════\n");

    // ---------------------------------------------------------
    // STEP 1: PREPARE DATA
    // ---------------------------------------------------------
    println!("📊 Step 1: Generating Dataset (Synthetic Fraud Data)...");

    // Train: 2000 samples
    let (train_x_raw, train_y_raw) = get_fraud_data(2000);
    // Transforma Vec<f32> a Vec<Vec<f32>> compatible con Dataset
    // (Ya viene como Vec<Vec> desde get_fraud_data, solo necesitamos envolver targets)
    let train_y_vec: Vec<Vec<f32>> = train_y_raw.iter().map(|&y| vec![y]).collect();

    // Test: 500 samples
    let (test_x_raw, test_y_raw) = get_fraud_data(500);
    let test_y_vec: Vec<Vec<f32>> = test_y_raw.iter().map(|&y| vec![y]).collect();

    let train_dataset = Dataset::new(train_x_raw, train_y_vec, 32); // Batch size 32
    let test_dataset = Dataset::new(test_x_raw, test_y_vec, 32);

    println!("   Train samples: {}", train_dataset.inputs.len());
    println!("   Test samples:  {}", test_dataset.inputs.len());

    // ---------------------------------------------------------
    // STEP 2: DEFINE MODEL
    // ---------------------------------------------------------
    println!("\n🧠 Step 2: Defining MLP Architecture...");

    // Input 3 -> L1(16) -> L2(8) -> Out(1) -> Sigmoid
    let model = Sequential::new(vec![
        Box::new(Linear::new(3, 16)?),
        Box::new(ReLU),
        Box::new(Linear::new(16, 8)?),
        Box::new(ReLU),
        Box::new(Linear::new(8, 1)?),
        Box::new(Sigmoid), // Convierte logits a probabilidad [0, 1]
    ]);

    println!("   Architecture: Linear(3->16) -> ReLU -> Linear(16->8) -> ReLU -> Linear(8->1) -> Sigmoid");

    // ---------------------------------------------------------
    // STEP 3: TRAIN
    // ---------------------------------------------------------
    println!("\n🏋️  Step 3: Training with Custom BCELoss...");

    let mut trainer = TrainerBuilder::new(model)
        .learning_rate(0.01)
        .build_adam(Box::new(BCELoss)); // Using build_adam as build_adamw is not exposed yet

    let history = trainer.fit(&train_dataset, Some(&test_dataset), 10, true)?;

    let final_loss = history.last().unwrap().0.loss;
    println!("   ✓ Training finished. Final Loss: {:.4}", final_loss);

    // ---------------------------------------------------------
    // STEP 4: EXPORT TO ONNX
    // ---------------------------------------------------------
    println!("\n📦 Step 4: Exporting to ONNX...");

    // Trace with dummy input
    let dummy_input = Tensor::new(Array::zeros(vec![1, 3]), false);

    // Ensure eval mode (though MLP has no dropout/bn here, good practice)
    trainer.model_mut().eval();
    let output = trainer.model().forward(&dummy_input)?;

    let model_path = "fraud_model.onnx.json";
    numrs::ops::export::export_to_onnx(&output, model_path)?;
    println!("   -> Saved to '{}'", model_path);

    // ---------------------------------------------------------
    // STEP 5: PRODUCTION INFERENCE
    // ---------------------------------------------------------
    println!("\n🚀 Step 5: Simulating Production Inference...\n");

    if !Path::new(model_path).exists() {
        return Err(anyhow::anyhow!("Model file not found"));
    }

    // 1. Load Model (Simulate server startup)
    let onnx_model = load_onnx(model_path)?;
    println!("   [Server] Model loaded successfully.");

    // 2. Simulate incoming requests
    let cases = vec![
        // Amount, Hour, Risk (Expected)
        (0.1, 0.5, 0.1, "Legit"),  // Low risk, low amount
        (0.9, 0.2, 0.96, "Fraud"), // High risk (>0.95 condition)
        (0.6, 0.8, 0.85, "Fraud"), // High risk (>0.8) AND high amount (>0.5)
        (0.2, 0.9, 0.85, "Legit"), // High risk BUT low amount (Should be legit per logic)
    ];

    println!("   Processing transactions:");
    println!("   ---------------------------------------------------------------");
    println!(
        "   | {:<6} | {:<6} | {:<6} | {:<8} | {:<8} | {:<6} |",
        "Amount", "Hour", "Risk", "Prob", "Pred", "Match?"
    );
    println!("   ---------------------------------------------------------------");

    let input_name = &onnx_model.graph.inputs[0].name;

    for (amt, hr, rsk, label) in cases {
        // Prepare input map
        let input_arr = Array::new(vec![1, 3], vec![amt, hr, rsk]);
        let mut inputs = HashMap::new();
        inputs.insert(input_name.clone(), input_arr);

        // Run Inference
        let results = infer(&onnx_model, inputs)?;
        let output_tensor = results
            .get("tensor_0")
            .or_else(|| results.get("output"))
            .unwrap(); // Output name might vary, check graph usually

        let prob = output_tensor.data[0];
        let pred = if prob > 0.5 { "Fraud" } else { "Legit" };
        let match_ok = if pred == label { "✅" } else { "❌" };

        println!(
            "   | {:<6.2} | {:<6.2} | {:<6.2} | {:<8.4} | {:<8} |   {}    |",
            amt, hr, rsk, prob, pred, match_ok
        );
    }
    println!("   ---------------------------------------------------------------");

    Ok(())
}
